### PR TITLE
Fixed processing order for update/ get events in the extension

### PR DESCRIPTION
**Pull request by swarna04**
## Description

Ref: [MOB-19606]

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/AEPOptimize/Optimize.swift [hunk 1]
```diff
@@ -82,6 +94,23 @@ public class Optimize: NSObject, Extension {
...
+        registerListener(type: EventType.optimize,
+                         source: EventSource.contentComplete,
+                         listener: processUpdatePropositionsCompleted(event:))
+
+        // Handler function called for each queued event. If the queued event is a get propositions event, process it
+        // otherwise if it is an Edge event to update propositions, process it only if it is completed.
+        eventsQueue.setHandler { event -> Bool in
+            if event.isGetEvent {
+                self.processGetPropositions(event: event)
+            } else if event.type == EventType.edge {
```
Comment (sbenedicadb) [inline Sources/AEPOptimize/Optimize.swift:108]: use your extension?
```
} else if event.isUpdateEvent {
```

File: Sources/AEPOptimize/Optimize.swift [hunk 2]
```diff
@@ -116,11 +145,10 @@ public class Optimize: NSObject, Extension {
             return
         }
 
-        let targetDecisionScopes = decisionScopes
+        let validDecisionScopes = decisionScopes
             .filter { $0.isValid }
-            .compactMap { $0.name }
 
-        if targetDecisionScopes.isEmpty {
+        guard !validDecisionScopes.isEmpty else {
```
Comment (sbenedicadb) [inline Sources/AEPOptimize/Optimize.swift:151]: 👍

File: Sources/AEPOptimize/Optimize.swift [hunk 3]
```diff
@@ -149,16 +177,93 @@ public class Optimize: NSObject, Extension {
...
+                                                                       ])
+            self.dispatch(event: updateCompleteEvent)
+        }
+
+        // In AEP Response Event handle, `requestEventId` corresponds to the UUID for the Edge request.
+        // Storing the request event UUID to compare and process only the anticipated response in the extension.
+        updateRequestEventIdsInProgress[edgeEvent.id.uuidString] = validDecisionScopes
+
+        // add the Edge event to update propositions in the events queue.
+        eventsQueue.add(edgeEvent)
```
Comment (sbenedicadb) [inline Sources/AEPOptimize/Optimize.swift:221]: makes more sense to the reader to move these above the `MobileCore.dispatch` call i think.

particularly, i notice manipulating the `updateRequestEventIdsInProgress` dictionary in the callback, but we hadn't seen any entry made to that dictionary at that point.

**Review by sbenedicadb — APPROVED** [review]
looks good - couple of small nits

File: README.md [hunk 4]
```diff
@@ -49,7 +49,7 @@ Alternatively, if your project has a `Package.swift` file, you can add AEPOptimi
 
 ```
 dependencies: [
-    .package(url: "https://github.com/adobe/aepsdk-optimize-ios.git", .upToNextMajor(from: "4.0.0"))
+    .package(url: "https://github.com/adobe/aepsdk-optimize-ios.git", .upToNextMajor(from: "4.0.2"))
```
[the remark this answers is not in the document]
Comment (swarna04) [reply README.md:52]: Yup, just making sure the value here reflects the latest version!

File: Sources/AEPOptimize/Optimize.swift (hunk 1 above)
Comment (swarna04) [reply Sources/AEPOptimize/Optimize.swift:108]: Nope, we are tracking the event-request content dispatched to edge extension (as a result of update propositions API call). The content complete response event will contain the edge event's ID for parent ID.

File: Sources/AEPOptimize/Optimize.swift (hunk 3 above)
Comment (swarna04) [reply Sources/AEPOptimize/Optimize.swift:221]: Yeah, can move those above but I believe these will almost always be executed before the async callback is invoked.

**Pull request merged**